src/components/Main/Main.jsx

```diff
@@ -1,15 +1,202 @@
-import React from 'react';
+import React, { useEffect, useState } from 'react';
 import { bool, shape } from 'prop-types';
 
 import connectToDatoCms from '@root/hoc/connectToDatoCms';
 
 import './Main.scss';
+import { map } from 'lodash-es';
 
 const Main = ({ client, devMode, plugin }) => {
   if (devMode) {
     console.debug({ client, plugin }); // eslint-disable-line no-console
   }
-  return <div className="container">Your Plugin goes here</div>;
+
+  let [source, setSource] = useState('');
+  let [name, setName] = useState('');
+  let [url, setUrl] = useState('');
+  let [imgLst, setImgLst] = useState([]);
+  let [selected, setSelected] = useState([]);
+  let [doneLoading, setDoneLoading] = useState(false);
+
+  const get_model = async (id) => {
+    try {
+      let ret = await client.items.find(id, {});
+      return ret;
+    } catch (err) {
+      return {};
+    }
+  };
+
+  const parse_nested_content = async (obj) => {
+    let ids = [];
+    for (const property in obj) {
+      if (typeof obj[property] === 'string' && property !== 'id') {
+        // this could be a linked instance
+        let nested = await get_model(obj[property]);
+
+        if (nested) {
+          ids = ids.concat(await parse_nested_content(nested));
+        }
+      } else {
+        let nested = obj[property];
+        if (nested) {
+          if (nested.constructor === Array) {
+            // this could be multiple assets
+            nested.forEach((asset) => {
+              if (asset.uploadId) {
+                ids.push(asset.uploadId);
+              }
+            });
+          } else {
+            if (nested.uploadId) {
+              ids.push(nested.uplpadId);
+            }
+          }
+        }
+      }
+    }
+
+    return ids;
+  };
+
+  const parse_field = async (field) => {
+    let ids = [];
+    if (field.field_type === 'file') {
+      ids.push(plugin.getFieldValue(field.api_key).upload_id);
+    } else {
+      // this is a link to a model instance
+      let itemId = plugin.getFieldValue(field.api_key);
+      ids = ids.concat(await parse_nested_content(await client.items.find(itemId)));
+    }
+    return ids;
+  };
```

Comment:
Why create array here? 

```suggestion
const parse_field = async ({ field_type: fieldType, api_key: apiKey }) => {
  switch(fieldType) {
    case 'file':
      return plugin.getFieldValue(field.api_key).upload_id;
    case default:
      const itemId = plugin.getFieldValue(apiKey);
      return await parse_nested_content(await client.items.find(itemId))
  }
}
```